server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java

```diff
@@ -0,0 +1,488 @@
+                .map(AsyncBatchShardFetch.NodeEntry::getNodeId)
+                .map(nodes::get)
+                .toArray(DiscoveryNode[]::new);
+            asyncFetch(discoNodesToFetch, fetchingRound);
+        }
+
+        // if we are still fetching, return null to indicate it
+        if (hasAnyNodeFetching(cache)) {
+            return new AsyncBatchShardFetch.FetchResult<>(null, emptyMap());
+        } else {
+            // nothing to fetch, yay, build the return value
+            Map<DiscoveryNode, T> fetchData = new HashMap<>();
+            Set<String> failedNodes = new HashSet<>();
+            for (Iterator<Map.Entry<String, AsyncBatchShardFetch.NodeEntry<T>>> it = cache.entrySet().iterator(); it.hasNext();) {
+                Map.Entry<String, AsyncBatchShardFetch.NodeEntry<T>> entry = it.next();
+                String nodeId = entry.getKey();
+                AsyncBatchShardFetch.NodeEntry<T> nodeEntry = entry.getValue();
+
+                DiscoveryNode node = nodes.get(nodeId);
+                if (node != null) {
+                    if (nodeEntry.isFailed()) {
+                        // if its failed, remove it from the list of nodes, so if this run doesn't work
+                        // we try again next round to fetch it again
+                        it.remove();
+                        failedNodes.add(nodeEntry.getNodeId());
+                    } else {
+                        if (nodeEntry.getValue() != null) {
+                            fetchData.put(node, nodeEntry.getValue());
+                        }
+                    }
+                }
+            }
+            Set<String> allIgnoreNodes = unmodifiableSet(new HashSet<>(nodesToIgnore));
+            // clear the nodes to ignore, we had a successful run in fetching everything we can
+            // we need to try them if another full run is needed
+            nodesToIgnore.clear();
+
+            // if at least one node failed, make sure to have a protective reroute
+            // here, just case this round won't find anything, and we need to retry fetching data
+            if (failedNodes.isEmpty() == false || allIgnoreNodes.isEmpty() == false) {
+                reroute(batchUUID, "nodes failed [" + failedNodes.size() + "], ignored [" + allIgnoreNodes.size() + "]");
+            }
+
+            return new AsyncBatchShardFetch.FetchResult<>(fetchData, ignoredShardToNodes);
+        }
+    }
+
+    /**
+     * Called by the response handler of the async action to fetch data. Verifies that its still working
+     * on the same cache generation, otherwise the results are discarded. It then goes and fills the relevant data for
+     * the shard (response + failures), issuing a reroute at the end of it to make sure there will be another round
+     * of allocations taking this new data into account.
+     */
+    protected synchronized void processAsyncFetch(List<T> responses, List<FailedNodeException> failures, long fetchingRound) {
```

Comment:
Completely same as `AsyncShardFetch` with only difference instead of using ShardId in logging we are using batchId